File: src/main/java/com/kewebsi/util/KewebsiDateUtils.java
package com.kewebsi.util;

import com.kewebsi.errorhandling.CodingErrorException;
import com.kewebsi.errorhandling.StringParsingError;
import com.kewebsi.html.dateeditor.CalendarController;
import com.kewebsi.html.dateeditor.StringParsingResult;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

public class KewebsiDateUtils {
    /**
     * Parses a date string in one of the forms:
     * day-month-year, // German
     * month-day-year, // American
     * year-month-day, // Technical/sortable
     * year-day-month, // Uncommon
     * day-month, // The year is then assumed the current year.
     * month-day // The year is then assumed the current year.
     * The separator is any sequence of non-digit characters.

     */
    public static LocalDate parseDate(String dateStr, boolean monthBeforeDay, boolean yearFirst, CalendarController.YEAR_DIGITS yearDigits) throws StringParsingError {


        StringParsingResult firstIntResult = getNextInt(dateStr, 0);

        if (firstIntResult.idx < 0) {
            throw new StringParsingError("Invalid data format", dateStr);
        }

        StringParsingResult secondIntResult = getNextInt(dateStr, firstIntResult.idx);

        if (secondIntResult.idx < 0) {
            throw new StringParsingError("Invalid data format", dateStr);
        }

        StringParsingResult thirdIntResult = getNextInt(dateStr, secondIntResult.idx);

        boolean yearProvided = true;
        if (thirdIntResult.idx < 0) {
            yearProvided = false;
        }

        if (! yearProvided && yearDigits== CalendarController.YEAR_DIGITS.Required) {
            throw new StringParsingError("Invalid data format. Year value must be provided", dateStr);
        }

        if (yearProvided && yearDigits== CalendarController.YEAR_DIGITS.Forbidden) {
            throw new StringParsingError("Invalid data format. Unexpected year value provided", dateStr);
        }

        int year;
        int month;
        int day;

        if (yearProvided) {
            if (yearFirst) {
                year = firstIntResult.result;
                if (monthBeforeDay) {
                    month = secondIntResult.result;
                    day = thirdIntResult.result;
                } else {
                    day = secondIntResult.result;
                    month = thirdIntResult.result;
                }
            } else {
                if (monthBeforeDay) {
                    month = firstIntResult.result;
                    day = secondIntResult.result;
                } else {
                    day = firstIntResult.result;
                    month = secondIntResult.result;
                }
                year = thirdIntResult.result;
            }
        } else {
            year = LocalDate.now().getYear();
            if (monthBeforeDay) {
                month = firstIntResult.result;
                day = secondIntResult.result;
            } else {
                day = firstIntResult.result;
                month = secondIntResult.result;
            }
        }

        LocalDate result;
        try {
            result = LocalDate.of(year, month, day);
        } catch (Exception e) {
            throw new StringParsingError(e);
        }
        return result;
    }


    public static LocalTime parseTime(String timeStr) throws StringParsingError {
        return parseTime(timeStr, 0);
    }

    public static LocalTime parseTime(String timeStr, int startIdx) throws StringParsingError {
        StringParsingResult hours = getNextInt(timeStr, startIdx);

        if (hours.idx < 0) {
            throw new StringParsingError("Invalid data format", timeStr);
        }

        StringParsingResult minutes = getNextInt(timeStr, hours.idx);

        if (minutes.idx < 0) {
            throw new StringParsingError("Invalid data format", timeStr);
        }

        int amOptions = 0;  // 0 = no AM/PM, 1 = AM, 2 = PM

        String timeStringUpperCase = timeStr.toUpperCase();
        if (timeStringUpperCase.contains("AM")) {
            amOptions = 1;
        } else {
            if (timeStringUpperCase.contains("PM")) {
                amOptions = 2;
            }
        }

        int hh = hours.result;
        int mm = minutes.result;

        if (amOptions > 0 && hours.result > 12) {
            throw new StringParsingError("Do not specify hours > 12 when specifying AM or PM ", String.valueOf(hh));
        }

        if (amOptions == 2) {
            hh +=12;
        }

        return LocalTime.of(hh, mm);

    }

    public static LocalDateTime parseDateTime(String dateTimeStr, boolean monthBeforeDay, boolean yearFirst, CalendarController.YEAR_DIGITS yearDigits) throws StringParsingError {

        if (yearDigits == CalendarController.YEAR_DIGITS.Optional) {
            throw new CodingErrorException(yearDigits + " not supported here.");
        }

        StringParsingResult firstIntResult = getNextInt(dateTimeStr, 0);

        if (firstIntResult.idx < 0) {
            throw new StringParsingError("Invalid data format", dateTimeStr);
        }

        StringParsingResult secondIntResult = getNextInt(dateTimeStr, firstIntResult.idx);

        if (secondIntResult.idx < 0) {
            throw new StringParsingError("Invalid data format", dateTimeStr);
        }

        StringParsingResult thirdIntResult = getNextInt(dateTimeStr, secondIntResult.idx);

        if (thirdIntResult.idx < 0) {
            throw new StringParsingError("Invalid data format", dateTimeStr);
        }

        StringParsingResult fourthIntResult = getNextInt(dateTimeStr, thirdIntResult.idx);

        if (fourthIntResult.idx < 0) {
            throw new StringParsingError("Invalid data format", dateTimeStr);
        }

        StringParsingResult fifthIntResult = null;

        if (yearDigits == CalendarController.YEAR_DIGITS.Required) {

            fifthIntResult = getNextInt(dateTimeStr, fourthIntResult.idx);
            if (fifthIntResult.idx < 0) {
                throw new StringParsingError("Invalid data format", dateTimeStr);
            }
        }


        boolean yearProvided = (yearDigits == CalendarController.YEAR_DIGITS.Required);

        int year;
        int month;
        int day;

        if (yearProvided) {
            if (yearFirst) {
                year = firstIntResult.result;

                // Date calculations in the database can crash due to a limited number range.
                // Hence we restrict date values to 4 digits.
                if (year < -9999 || year > 9999) {
                    throw new StringParsingError("Invalid data format. Year out of range", dateTimeStr);
                }

                if (monthBeforeDay) {
                    month = secondIntResult.result;
                    day = thirdIntResult.result;
                } else {
                    day = secondIntResult.result;
                    month = thirdIntResult.result;
                }
            } else {
                if (monthBeforeDay) {
                    month = firstIntResult.result;
                    day = secondIntResult.result;
                } else {
                    day = firstIntResult.result;
                    month = secondIntResult.result;
                }
                year = thirdIntResult.result;
            }
        } else {
            year = LocalDate.now().getYear();
            if (monthBeforeDay) {
                month = firstIntResult.result;
                day = secondIntResult.result;
            } else {
                day = firstIntResult.result;
                month = secondIntResult.result;
            }
        }

        int hour;
        int minute;
        if (yearProvided) {
            hour = fourthIntResult.result;
            minute = fifthIntResult.result;;
        } else {
            hour = thirdIntResult.result;
            minute = fourthIntResult.result;
        }

        LocalDateTime result;
        try {
            result = LocalDateTime.of(year, month, day,hour,minute);
        } catch (Exception e) {
            throw new StringParsingError(e);
        }
        return result;
    }


    public static LocalDate parseGermanDate(String dateStr) throws StringParsingError {
        LocalDate result = KewebsiDateUtils.parseDate(dateStr.trim(), false, false, CalendarController.YEAR_DIGITS.Required);
        return result;
    }


    public static LocalDateTime parseGermanDateTime(String dateStr) throws StringParsingError {
        LocalDateTime result = KewebsiDateUtils.parseDateTime(dateStr.trim(), false, false, CalendarController.YEAR_DIGITS.Required);
        return result;
    }

    public static String printLocalDate(LocalDate localDate, boolean monthBeforeDay, boolean yearFirst, int yearDigits, String divider) {
        String yearStr = String.valueOf(localDate.getYear());
        int length = yearStr.length();

        if (yearDigits <=2) {
            yearStr = yearStr.substring(length - 2, length);
        }

        String monthStr = String.valueOf(localDate.getMonthValue());
        String dayStr = String.valueOf(localDate.getDayOfMonth());


        String result = getStringForYearMonthDay(yearStr, monthStr, dayStr, monthBeforeDay, yearFirst, divider);

        return result;
    }


    public static String printLocalDateTime(LocalDateTime localDate, boolean monthBeforeDay, boolean yearFirst, int yearDigits, boolean printSeconds, boolean printNanoSeconds, String divider) {
        // TODO SPEEDUP: String operations can be drastically simplified.
        String yearStr = String.valueOf(localDate.getYear());
        int length = yearStr.length();

        if (yearDigits <=2) {
            yearStr = yearStr.substring(length - 2, length);
        }

        String monthStr = String.valueOf(localDate.getMonthValue());
        String dayStr = String.valueOf(localDate.getDayOfMonth());
        String hourStr = String.valueOf(localDate.getHour());
        String minuteStr = String.valueOf(localDate.getMinute());
        String secondStr = String.valueOf(localDate.getSecond());
        String nanoSecondStr = String.valueOf(localDate.getNano());



        String result = getStringForYearMonthDay(yearStr, monthStr, dayStr, monthBeforeDay, yearFirst, divider);

        hourStr =  CommonUtils.fillWithLeadingZeros(2, hourStr);
        minuteStr =  CommonUtils.fillWithLeadingZeros(2, minuteStr);
        secondStr =  CommonUtils.fillWithLeadingZeros(2, secondStr);
        nanoSecondStr =  CommonUtils.fillWithLeadingZeros(9, nanoSecondStr);

        result += " " + hourStr + ":" + minuteStr;

        if (printSeconds) {
            result += ":" + secondStr;
            if (printNanoSeconds) {
                result += "." + nanoSecondStr;
            }
        }

        return result;
    }

    private static String getStringForYearMonthDay(String yearStr, String monthStr, String dayStr, boolean monthBeforeDay, boolean yearFirst, String divider) {
        String monthStringTwoDigits = CommonUtils.fillWithLeadingZeros(2, monthStr);
        String dayStringTwoDigits = CommonUtils.fillWithLeadingZeros(2, dayStr);


        String result;
        if (yearFirst) {
            if (monthBeforeDay) {
                result = yearStr + divider + monthStr + divider + dayStr;
            } else {
                result = yearStr + divider + dayStringTwoDigits + divider + monthStringTwoDigits;
            }

        } else {
            if (monthBeforeDay) {
                result = monthStr + divider + dayStr + divider + yearStr;
            } else {
                result = dayStringTwoDigits + divider + monthStringTwoDigits + divider + yearStr;
            }
        }
        return result;
    }


//    public static String printDateTime(LocalDateTime localDate, boolean monthBeforeDay, boolean yearFirst, int yearDigits, String divider) {
//        String res
//    }

    public static String printGermanDate(LocalDate localDate) {
        return printLocalDate(localDate, false, false, 4, ".");
    }

    public static String printGermanDateTime(LocalDateTime localDate) {
        return printLocalDateTime(localDate, false, false, 4, true, true,".");
    }

    public static String printTime24h(LocalTime localTime) {
        return CommonUtils.twoDigits(localTime.getHour()) + ":" + CommonUtils.twoDigits(localTime.getMinute());
    }


    /**
     * We start at the beginning of the integer or at a non-digit in front of the integer.
     * @param dateStr
     * @param idx
     * @return
     */
    public static StringParsingResult getNextInt(String dateStr, int idx) {
        String outStr = "";
        boolean digitStarted = false;
        boolean digitCompleted = false;
        while (idx < dateStr.length() && !digitCompleted) {
            char run = dateStr.charAt(idx);
            if (isDigit(run)) {
                digitStarted = true;
                outStr += run;
            } else {
                if (digitStarted) {
                    digitCompleted = true;
                }
            }

            if (idx == dateStr.length()-1 && digitStarted) {
                digitCompleted = true;
            }

            idx ++;
        }
        StringParsingResult result = null;
        if (!digitCompleted) {
            result = new StringParsingResult(-1 ,-1);
        } else {
            int outInt = Integer.parseInt(outStr);
            result = new StringParsingResult(idx ,outInt);
        }
        return result;
    }

    public static boolean isDigit(char c) {
        if (c >='0' && c <='9') {
            return true;
        }
        return false;
    }


}

File: src/main/resources/webFrontend/scripts/ts/modules/kewebsiDateUtils.ts
import {LocalDate, LocalTime, LocalDateTime, createLocalDateTime} from "./dateTime.js"
import { DateTimeWire, DateTimeWireOrNull } from "./jointTypes.js";
import {fillWithLeadingZeros, isNullOrUndefined, twoDigits} from "./stringUtils.js"


type StringParsingResult = {
    idx: number;
    result : number;
}

type ParsingError = {
    errorMessage: string
    errorValue?: string
}

export function isParsingError<T extends object>( locatDateOrParsingError: T | ParsingError) : locatDateOrParsingError is ParsingError {
    return 'errorMessage' in locatDateOrParsingError;
}

const YEAR_REQUIRED = "YEAR_REQUIRED";
const YEAR_FORBIDDEN = "YEAR_FORBIDDEN";
const YEAR_OPTIONAL = "YEAR_OPTIONAL";



/**
 * Parses a date string in one of the forms:
 * day-month-year, // German
 * month-day-year, // American
 * year-month-day, // Technical/sortable
 * year-day-month, // Uncommon
 * day-month, // The year is then assumed the current year.
 * month-day // The year is then assumed the current year.
 * The separator is any sequence of non-digit characters.

    */
export function parseDate(dateStr: string, monthBeforeDay: boolean, yearFirst: boolean, yearDigitsMode: string) : (LocalDate | ParsingError) {


    const firstIntResult : StringParsingResult = getNextInt(dateStr, 0);

    if (firstIntResult.idx < 0) {
        return {errorMessage: "Invalid date format: ", errorValue:  dateStr};
    }

    const secondIntResult: StringParsingResult = getNextInt(dateStr, firstIntResult.idx);

    if (secondIntResult.idx < 0) {
        return {errorMessage: "Invalid date format. ", errorValue:  dateStr};
    }

    const thirdIntResult: StringParsingResult = getNextInt(dateStr, secondIntResult.idx);

    const yearProvided = thirdIntResult.idx < 0 ? false : true;

    if (! yearProvided && yearDigitsMode === YEAR_REQUIRED) {
        return {errorMessage: "Invalid date format. Year value must be provided. "};
    }

    if (yearProvided && yearDigitsMode === YEAR_FORBIDDEN) {
        return {errorMessage: "Invalid date format. Unexpected year value provided."};
    }

    let year = 0;
    let month = 0;
    let day = 0;

    if (yearProvided) {
        if (yearFirst) {
            year = firstIntResult.result;
            if (monthBeforeDay) {
                month = secondIntResult.result;
                day = thirdIntResult.result;
            } else {
                day = secondIntResult.result;
                month = thirdIntResult.result;
            }
        } else {
            if (monthBeforeDay) {
                month = firstIntResult.result;
                day = secondIntResult.result;
            } else {
                day = firstIntResult.result;
                month = secondIntResult.result;
            }
            year = thirdIntResult.result;
        }
    } else {
        const year = new Date().getFullYear()
        if (monthBeforeDay) {
            month = firstIntResult.result;
            day = secondIntResult.result;
        } else {
            day = firstIntResult.result;
            month = secondIntResult.result;
        }
    }

    const resultDate : LocalDate = {year: year, month: month, day: day}; // Note that we do not check here, whether the date is possible. (E.g. 33.1.2020)
    return resultDate;
}


/**
 * Parses the string in ISO8601 format of the variation yyyy-mm-ddThh:mm:ss.f
 * @param dtstr 
 * @returns 
 */
export function parseIsoDateTime(dtstr: string) : LocalDateTime {
    let strArr: string[]  = dtstr.split(/[:. T-]/);

		
    let year = parseFloat(strArr[0]);;
    let month = 1;
    let day = 1
    let hour = 0;
    let minute = 0;
    let seconds = 0;
    let fraction = 0;
    
    if (strArr.length >= 2) {
    	month = parseFloat(strArr[1]);
    }
    if (strArr.length >= 3) {
    	day = parseFloat(strArr[2]);
    }
    if (strArr.length >= 4) {
    	hour = parseFloat(strArr[3]);
    }
    if (strArr.length >= 5) {
    	minute = parseFloat(strArr[4]);
    }
    if (strArr.length >= 6) {
    	seconds = parseFloat(strArr[5]);
    }
    if (strArr.length == 7) {
        let fractionStr = strArr[6];
        let fractionTop = parseFloat(fractionStr);
        let fractionBottom = Math.pow(10, fractionStr.length);
        fraction = fractionTop/fractionBottom;
    }

    let ldt = createLocalDateTime(year, month, day, hour, minute, seconds);
    
    return ldt;

}


export function  parseTime(timeStr: string) : (LocalTime | ParsingError) {
    return parseTimeStartingFromIdx(timeStr, 0);
}

function parseTimeStartingFromIdx(timeStr: string, startIdx: number) : (LocalTime | ParsingError) {
    const hours : StringParsingResult  = getNextInt(timeStr, startIdx);

    if (hours.idx < 0) {
        return {errorMessage: "Invalid time format."};
    }

    const minutes : StringParsingResult  = getNextInt(timeStr, hours.idx);

    if (minutes.idx < 0) {
        return {errorMessage: "Invalid time format."};
    }



    let amOptions = 0;  // 0 = no AM/PM, 1 = AM, 2 = PM

    const timeStringUpperCase = timeStr.toUpperCase();
    if (timeStringUpperCase.includes("AM")) {
        amOptions = 1;
    } else {
        if (timeStringUpperCase.includes("PM")) {
            amOptions = 2;
        }
    }

    let hh = hours.result;
    const mm = minutes.result;

    if (amOptions > 0 && hours.result > 12) {
        return {errorMessage: "Invalid data format.", errorValue: hh.toString()};
    }

    if (amOptions == 2) {
        hh +=12;
    }

    if (mm < 0 || mm > 59) {
        return {errorMessage: "Invalid time format. Minutes out of range."};
    }

    let result: LocalTime = {hour: hh, minute: mm, second:0, nanosecond: 0 }
    return result;

}

export function parseDateTimeIgnoresHoursAndMinutesXXX(dateTimeStr: string, monthBeforeDay: boolean,  yearFirst: boolean, yearDigitsMode: string) : (LocalDateTime | ParsingError) {

    const firstIntResult : StringParsingResult = getNextInt(dateTimeStr, 0);

    if (firstIntResult.idx < 0) {
        return {errorMessage: "Invalid datetime format: ", errorValue:  dateTimeStr};
    }

    const secondIntResult: StringParsingResult = getNextInt(dateTimeStr, firstIntResult.idx);

    if (secondIntResult.idx < 0) {
        return {errorMessage: "Invalid datetime format. ", errorValue:  dateTimeStr};
    }

    const thirdIntResult: StringParsingResult = getNextInt(dateTimeStr, secondIntResult.idx);

    const yearProvided = thirdIntResult.idx < 0 ? false : true;

    if (! yearProvided && yearDigitsMode === "YEAR_REQUIRED") {
        return {errorMessage: "Invalid datetime format. Year value must be provided. "};
    }

    if (yearProvided && yearDigitsMode === "YEAR_FORBIDDEN") {
        return {errorMessage: "Invalid datetime format. Unexpected year value provided."};
    }

    let year = 0;
    let month = 0;
    let day = 0;

    if (yearProvided) {
        if (yearFirst) {
            year = firstIntResult.result;
            if (monthBeforeDay) {
                month = secondIntResult.result;
                day = thirdIntResult.result;
            } else {
                day = secondIntResult.result;
                month = thirdIntResult.result;
            }
        } else {
            if (monthBeforeDay) {
                month = firstIntResult.result;
                day = secondIntResult.result;
            } else {
                day = firstIntResult.result;
                month = secondIntResult.result;
            }
            year = thirdIntResult.result;
        }
    } else {
        const year = new Date().getFullYear()
        if (monthBeforeDay) {
            month = firstIntResult.result;
            day = secondIntResult.result;
        } else {
            day = firstIntResult.result;
            month = secondIntResult.result;
        }
    }

    const resultDate : LocalDate = {year: year, month: month, day: day}; // Note that we do not check here, whether the date is possible. (E.g. 33.1.2020)
    const resultTime: LocalTime = {hour: 0, minute: 0, second: 0, nanosecond: 0}; // Note that we do not check here, whether the time is possible. (E.g. 33:70h)
    const result:  LocalDateTime = { date: resultDate, time: resultTime } ;

    return result;
}


export function  parseGermanDate(dateStr: string) : (LocalDate | ParsingError) {
   return parseDate(dateStr.trim(), false, false, YEAR_REQUIRED);
}


export function parseGermanDateTime(dateStr: string) : (LocalDateTime | ParsingError) {
    return  parseDateTimeIgnoresHoursAndMinutesXXX(dateStr.trim(), false, false, YEAR_REQUIRED);
}

function printLocalDate(localDate: LocalDate, monthBeforeDay: boolean, yearFirst: boolean, yearDigits: number, divider: string) : string {
    let yearStr = localDate.year.toString();
    const length = yearStr.length;

    if (yearDigits <=2) {
        yearStr = yearStr.substring(length - 2, length);
    }

    const monthStr = localDate.month.toString();
    const dayStr = localDate.day.toString();

    const result = getStringForYearMonthDay(yearStr, monthStr, dayStr, monthBeforeDay, yearFirst, divider);

    return result;
}


function printLocalDateTime(localDate: LocalDateTime, monthBeforeDay: boolean, yearFirst: boolean, yearDigits: number, printSeconds: boolean, printNanoSeconds: boolean, divider: string) : string {
    // TODO SPEEDUP: String operations can be drastically simplified.
    let yearStr = localDate.date.year.toString()
    const length = yearStr.length;

    if (yearDigits <=2) {
        yearStr = yearStr.substring(length - 2, length);
    }

    let monthStr = localDate.date.month.toString();
    let dayStr = localDate.date.day.toString();
    let hourStr = localDate.time.hour.toString();
    let minuteStr = localDate.time.minute.toString();
    let secondStr = localDate.time.second.toString();
    let nanoSecondStr = localDate.time.nanosecond.toString();   



    let result = getStringForYearMonthDay(yearStr, monthStr, dayStr, monthBeforeDay, yearFirst, divider);

    hourStr =  fillWithLeadingZeros(2, hourStr);
    minuteStr =  fillWithLeadingZeros(2, minuteStr);
    secondStr =  fillWithLeadingZeros(2, secondStr);
    nanoSecondStr =  fillWithLeadingZeros(9, nanoSecondStr);

    result += " " + hourStr + ":" + minuteStr;

    if (printSeconds) {
        result += ":" + secondStr;
        if (printNanoSeconds) {
            result += "." + nanoSecondStr;
        }
    }

    return result;
}

function getStringForYearMonthDay(yearStr: string, monthStr: string, dayStr: string, monthBeforeDay: boolean, yearFirst: boolean, divider: string) : string {
    const monthStringTwoDigits = fillWithLeadingZeros(2, monthStr);
    const dayStringTwoDigits = fillWithLeadingZeros(2, dayStr);


    let result: string;
    if (yearFirst) {
        if (monthBeforeDay) {
            result = yearStr + divider + monthStr + divider + dayStr;
        } else {
            result = yearStr + divider + dayStringTwoDigits + divider + monthStringTwoDigits;
        }

    } else {
        if (monthBeforeDay) {
            result = monthStr + divider + dayStr + divider + yearStr;
        } else {
            result = dayStringTwoDigits + divider + monthStringTwoDigits + divider + yearStr;
        }
    }
    return result;
}


function printGermanDate(localDate: LocalDate) : string {
    return printLocalDate(localDate, false, false, 4, ".");
}

function printGermanDateTime(localDate: LocalDateTime) : string {
    return printLocalDateTime(localDate, false, false, 4, true, true,".");
}

function printTime24h(localTime: LocalTime) : string{
    return twoDigits(localTime.hour) + ":" + twoDigits(localTime.minute);
}


/**
 * We start at the beginning of the integer or at a non-digit in front of the integer.
 * @param dateStr
 * @param idx
 * @return
 */
function  getNextInt(dateStr: string, idx: number) : StringParsingResult {
    let  outStr = "";
    let digitStarted = false;
    let digitCompleted = false;
    while (idx < dateStr.length && !digitCompleted) {
        let run = dateStr.charAt(idx);
        if (isDigit(run)) {
            digitStarted = true;
            outStr += run;
        } else {
            if (digitStarted) {
                digitCompleted = true;
            }
        }

        if (idx == dateStr.length-1 && digitStarted) {
            digitCompleted = true;
        }

        idx ++;
    }
    let result: StringParsingResult;
    if (!digitCompleted) {
        result = {idx: -1, result: -1};
    } else {
        let outInt = Number.parseInt(outStr);
        result = {idx: idx , result: outInt};
    }
    return result;
}

function isDigit(c: string) : boolean {
    if (c.length != 1) {
        throw Error("Unexpected string length. Expected length of 1 but got the following string: " + c);
    }

    const reg: RegExp = /\d/;

    return reg.test(c);
}

/**
 * 
 * @param dtwOrNull Must not be null. WireData is never null but might have nulls inside.
 * @returns 
 */
export function dateTimeWireToLocalDateTime(dtwOrNull: DateTimeWireOrNull) : LocalDateTime {
    if (isNullOrUndefined(dtwOrNull.value)) {
        return null;
    }
    const dtw = dtwOrNull.value;
    const localDate : LocalDate = {year: dtw.y, month: dtw.mo, day: dtw.d};
    const localTime : LocalTime = {hour: dtw.h, minute: dtw.mi, second: dtw.s, nanosecond: dtw.n};
    const localDateTime: LocalDateTime = {date: localDate, time: localTime}; 
    return localDateTime;
}

export function  localDateTimeToDateTimeWireOrNull(localDateTime: LocalDateTime ) : DateTimeWireOrNull {
    
    
    let dateTimeWire = localDateTime === null ? null :  {y: localDateTime.date.year, mo: localDateTime.date.month, d: localDateTime.date.day,
        h: localDateTime.time.hour, mi: localDateTime.time.minute, s: localDateTime.time.second, n: localDateTime.time.nanosecond }
    let result : DateTimeWireOrNull = {value: dateTimeWire}
    return result;;
}